cleanupNoncurrentVersions.js

```diff
@@ -0,0 +1,439 @@
+        String.fromCharCode(versionId.charCodeAt(versionId.length - 1) - 1);
+}
+
+function _triggerDeletesOnEligibleObjects(bucket, versions, deleteMarkers,
+                                          endOfListing, cb) {
+    const versionsToDelete = [];
+    versions.forEach(version => {
+        if (version.IsLatest !== false) {
+            nSkippedCurrent += 1;
+        } else if (!_lastModifiedIsEligible(version.LastModified)) {
+            nSkippedTooRecent += 1;
+        } else {
+            versionsToDelete.push({
+                Key: version.Key,
+                VersionId: version.VersionId,
+            });
+        }
+    });
+    let ret = null;
+    deleteMarkers.forEach(deleteMarker => {
+        if (!_lastModifiedIsEligible(deleteMarker.LastModified)) {
+            nSkippedTooRecent += 1;
+        } else {
+            // To avoid making existing noncurrent versions current,
+            // we must be cautious on which delete markers we may
+            // safely delete.
+            //
+            // We can delete a DM safely iff:
+            //
+            //   it is a noncurrent DM
+            // OR
+            //   at least one of the last listed version or DM key is
+            //   strictly greater than the current DM key to delete
+            // OR
+            //   we reached the end of the listing
+            //
+            // Otherwise, we must not remove the DM before further
+            // listing to make sure we also remove the associated
+            // noncurrent versions. In such case, we continue listing
+            // from that DM key and versionId at next iteration,
+            // instead of the returned next key/versionId markers.
+            if (deleteMarker.IsLatest === false
+                || (deleteMarkers[deleteMarkers.length - 1].Key
+                    > deleteMarker.Key)
+                || (versions.length > 0
+                    && (versions[versions.length - 1].Key
+                        > deleteMarker.Key))
+                || endOfListing) {
+                versionsToDelete.push({
+                    Key: deleteMarker.Key,
+                    VersionId: deleteMarker.VersionId,
+                });
+            } else {
+                ret = {
+                    NextKeyMarker: deleteMarker.Key,
+                    NextVersionIdMarker: decVersionId(deleteMarker.VersionId),
+                };
+            }
+        }
+    });
+    _triggerDeletes(bucket, versionsToDelete, cb);
+    return ret;
+}
+
+function _waitForDeletesCompletion(cb) {
+    if (batchDeleteInProgress) {
+        batchDeleteOnFullDrain = cb;
+        return undefined;
+    }
+    return process.nextTick(cb);
+}
+
+function triggerDeletesOnBucket(bucketName, cb) {
+    const bucket = bucketName.trim();
+    bucketInProgress = bucket;
+    log.info(`starting task for bucket: ${bucket}`);
+    let NextKeyMarker = null;
+    let NextVersionIdMarker = null;
+    if (KEY_MARKER || VERSION_ID_MARKER) {
```

Comment:
It seems that the key marker and delete id marker should be linked somehow to the bucket they belong to, as here they are read/written as global variables by the `triggerDeletesOnBucket` call on the first bucket even if they in fact belonged to the second bucket in the list (the use case for this would be someone resuming an interrupted run with the same command and "forgetting" to drop the buckets already done from the command line).

It would be a valid requirement IMO that they explicitly apply only to the first bucket listed (or that only one bucket is provided if they are present), in which case mentioning it in the documentation or checking that there's only one bucket argument would be enough.

Reply:
Indeed, not fond of this, maybe instead something that could work better is providing a single env var to pass containing the triplet bucket-key-version, although handling properly delimiters may be challenging (like internally we use `\0` for separating key and version, may not be appropriate since env vars cannot contain `\0` IIRC)

Reply:
I'll likely keep it as is unless I have a "eureka" moment later!

Reply:
Actually we could use URL encoding for all, and encode the separators as some non-encoded character

Reply:
@RachedBenMustapha Did the change, now we can keep the same bucket list and a single `MARKER='bucket|key|version'` env var.

I also renamed ENDPOINT to S3_ENDPOINT to have a cleaner integration with the orchestration script that will also need BUCKETD_ENDPOINT and SPROXYD_ENDPOINT.

Reply:
Nice approach